### PR TITLE
2232-3

### PR DESCRIPTION
**Pull request by ZhenyaZhak**
## Номер
2232

## Номер задания
3

## Ссылка на видео с демо
https://youtu.be/A38rSEd543w

## Комментарии
Выполнены все условия.

### CODE REVIEW

File: 2232/3/controllers/articles_controller.rb [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  end
+
+  # add new article
+  get '/new' do
+    erb :new
+  end
+
+  # create a new article
+  post '/new' do
+    obj = ArticleHelper.new(params[:link])
```
Comment (stahor) [inline 2232/3/controllers/articles_controller.rb]: obj?

File: 2232/3/controllers/articles_controller.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    @article = Article.create title: obj.title, link: params[:link], rating: obj.rating
```
Comment (stahor) [inline 2232/3/controllers/articles_controller.rb]: parentheses

File: 2232/3/controllers/articles_controller.rb (hunk 1 above)
Comment (ZhenyaZhak) [reply 2232/3/controllers/articles_controller.rb]: Sorry, I forgot about it

File: 2232/3/helpers/article_helper.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+class ArticleHelper
+  include SentimentParser
+
+  LIMIT_COMMENTS = 50
+  attr_reader :link, :comments
+
+  def initialize(link)
+    @link = link
+    @comments = create_comments
```
Comment (stahor) [inline 2232/3/helpers/article_helper.rb]: better use method with memoization

File: 2232/3/helpers/article_helper.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def title
+    page.title
+  end
+
+  def rating
+    comments.map(&:rating).reduce(:+) / comments.size
```
Comment (stahor) [inline 2232/3/helpers/article_helper.rb]: try `comments.sum(&:rating)`

File: 2232/3/helpers/sentiment_parser.rb [hunk 5]
```diff
@@ -0,0 +1,29 @@
+require 'uri'
+require 'net/https'
+
+module SentimentParser
+  ACCESS_KEY = '***'.freeze
```
Comment (stahor) [inline 2232/3/helpers/sentiment_parser.rb]: Use `dotenv` or `Figaro`

File: 2232/3/helpers/sentiment_parser.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
+require 'uri'
+require 'net/https'
+
+module SentimentParser
```
Comment (stahor) [inline 2232/3/helpers/sentiment_parser.rb]: why is this module? 
Use service objects

File: 2232/3/helpers/sentiment_parser.rb [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    response = Net::HTTP.start(URL.host, URL.port, use_ssl: URL.scheme == 'https') do |http|
+      http.request(create_request(comments))
+    end
+    JSON(response.body)['documents']
+  end
+
+  private
+
+  def create_request(comments)
+    request = Net::HTTP::Post.new(URL, 'Content-Type' => 'application/json', 'Ocp-Apim-Subscription-Key' => ACCESS_KEY)
```
Comment (stahor) [inline 2232/3/helpers/sentiment_parser.rb]: better to break this into multiple lines to make more readable

File: 2232/3/helpers/sentiment_parser.rb [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    request.body = documents(comments)
+    request
+  end
+
+  def documents(comments)
+    texts = comments.map do |author, text|
+      { 'id' => author, 'language' => 'ru', 'text' => text }
```
Comment (stahor) [inline 2232/3/helpers/sentiment_parser.rb]: and what if one author has multiple comments. Try to find other better id

File: 2232/3/helpers/article_helper.rb [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    "https://comments.api.onliner.by/news/#{category}.post/#{number}/comments?limit=9999"
+  end
+
+  def create_comments
+    comments = response.each_with_object({}) { |cmt, hash| hash[cmt['author']['name']] = cmt['text'].tr("\n", ' ') }
+    initialize_comments(comments, create_sentiments(comments))
+  end
+
+  def response
+    JSON.parse(Faraday.get(api_link).body)['comments'][0...LIMIT_COMMENTS]
```
Comment (stahor) [inline 2232/3/helpers/article_helper.rb]: `first(LIMIT_COMMENTS)`

File: 2232/3/helpers/sentiment_parser.rb (hunk 7 above)
Comment (stahor) [inline 2232/3/helpers/sentiment_parser.rb]: you have `Faraday`. Use it instead of `Net::HTTP`!

File: 2232/3/helpers/article_helper.rb [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+  end
+
+  def initialize_comments(comments, sentiments)
+    sentiments.each_with_object([]) do |sentiment, arr|
```
Comment (stahor) [inline 2232/3/helpers/article_helper.rb]: use map

File: 2232/3/helpers/article_helper.rb [hunk 11]
```diff
@@ -0,0 +1,50 @@
+require_relative 'sentiment_parser'
+require_relative 'comment_helper'
+
+class ArticleHelper
```
Comment (stahor) [inline 2232/3/helpers/article_helper.rb]: Why helper?
Describe the responsibility of this class.

File: 2232/3/models/comment.rb [hunk 12]
```diff
@@ -0,0 +1,5 @@
+class Comment < Ohm::Model
+  attribute :text
+  attribute :author
+  attribute :rating
+end
```
Comment (stahor) [inline 2232/3/models/comment.rb]: don't u want to add  a reference to the article

File: 2232/3/views/articles/articles.erb [hunk 13]
```diff
@@ -0,0 +1,24 @@
+<div class="row justify-content-md-center">
+  <h1>Articles</h1>
+</div>
+<table>
+  <tbody>
+    <% @articles.each do |article| %>
+      <tr>
+        <div class="list-group">
+          <td>
+            <a method="GET" href='/articles/<%= article.id %>' class="list-group-item list-group-item-action"><%= article.title %>
```
Comment (stahor) [inline 2232/3/views/articles/articles.erb]: to helper method

File: 2232/3/models/comment.rb (hunk 12 above)
Comment (ZhenyaZhak) [reply 2232/3/models/comment.rb]: The link to the article is stored in the Article module

File: 2232/3/helpers/article_helper.rb (hunk 9 above)
Comment (stahor) [inline 2232/3/helpers/article_helper.rb]: ok I don't want jsut first 50 comments. I want to see 50 most popular comments.

File: 2232/3/helpers/comment_helper.rb [hunk 14]
```diff
@@ -0,0 +1,9 @@
+class CommentHelper
+  attr_reader :text, :rating, :author
+
+  def initialize(author, text, rating)
```
Comment (stahor) [inline 2232/3/helpers/comment_helper.rb]: ???
Do you really need this class?

**Comment by ZhenyaZhak** [comment]
@stahor I took into account almost all the advice. Instead of dotenv or figaro used a Yaml file that is added to .gitignore. The link to Article has not been added to the Comment module, because I do not quite understand why it is needed there (it already exists in the Article, if it's really necessary, I can add it). Added generation of URL from the module (not sure whether it is correct, but it works). In the rest all has corrected.

**Comment by ZhenyaZhak** [comment]
Sorry that one commit.

File: 2232/3/helpers/article_parser.rb [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+  def initialize(link)
+    @link = link
+  end
+
+  def title
+    page.title
+  end
+
+  def rating
+    comments.map(&:rating).reduce(:+) / comments.size
```
Comment (stahor) [inline 2232/3/helpers/article_parser.rb]: sum

File: 2232/3/helpers/comments_parser.rb [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  LIMIT_COMMENTS = 50
+
+  def initialize(page, link)
+    @page = page
+    @link = link
+  end
+
+  def call
+    sentiments.map do |sentiment|
+      author_text = comments[sentiment['id'].to_i]
```
Comment (stahor) [inline 2232/3/helpers/comments_parser.rb]: author?

**Pull request merged**